Bugfixes

Geolykt opened the pull request:
The Fire enchantment would not work as  intended as the BlockBreakEvent was passed for some reason before it was cancelled, makeing sure that the event would be passed last would make sure that it wouldn't be canceled after it was recieved.

THIS COMMIT _MAY_ HAVE UNINTEDED SIDE EFFECTS (edit: from what It looks like, it doesn't have any(

Geolykt committed:
Fixed the Fire Enchantment

The Fire enchantment would not work as  intended as the BlockBreakEvent was passed for some reason before it was cancelled, makeing sure that the event would be passed last would make sure that it wouldn't be canceled after it was recieved.

THIS COMMIT MAY HAVE UNINTEDED SIDE EFFECTS

Geolykt committed:
Prevent frozenstep and netherstep blocks from breaking

Currently there is unexpected behaviour which would occur when a frozenstep or netherstep spawned block would be destroyed by explosives (like tnt) or moved by pistons. This commit aims to change that.

Note: this code should be optimised, as it introduces lag at following rates:
 -Addional piston lag when retraction or extending a piston proprtional to the blocks spawned by the frozenstep and netherstep enchantments
-Addional lag on explosions proprtional to the blocks spawned by the frozenstep and netherstep enchantments multiplied by the blocks cleared in the explosion

This commit also bumps the version of the plugin in the plugin.yml file to 1.6.3 as new methods have been added as well as to make the pom.xml and the plugin.yml versions the same

rfrowe commented:
Hi @Geolykt, thanks for the code. Please fix up your indentation. It seems that some lines in your diff are indented with spaces and some with tabs. This repository is styled with 4-space indentations.

Geolykt commented:
yea, I will fix that now

Geolykt committed:
Fix indentation

Geolykt commented:
Alright, since I suppose this will lead to nowwhere, I closed this

rfrowe commented on Zenchantments/src/main/java/zedly/zenchantments/WatcherEnchant.java:
```diff
@@ -75,6 +84,88 @@ public void onBlockShred(BlockShredEvent evt) {
+    }
+    
+    /**
+     * This method returns whether a block is protected by the plugin and whether it should be considered <br>
```
This method does not adhere to the contract specified in this comment. You describe this method as a getter but it has side effects. I understand this is mentioned via the `remove` parameter doc but you should also mention it in the doc body.

You only use this method with `remove == true` once, so my recommendation is that you remove the removal functionality from this method. Then, your usage above:

```java
isProtectedBlock(block, true);
```

becomes:

```java
if (isProtectedBlock(block)) {
    NetherStep.netherstepLocs.remove(block);
    FrozenStep.frozenstepLocs.remove(block);
}
```

In the event that we need to do this elsewhere, I'd recommend that you just create a method `removeBlock` which encapsulates this functionality. Then your usage would be:

```java
if (isProtectedBlock(block)) {
    removeBlock(block);
}
```

Since my suggestion below changes your implementation from O(n) complexity to O(1), this refactor would yield better code with essentially no performance loss.

rfrowe reviewed (commented):
I understand that this PR has been open for a while, but I'm not willing to merge code with TODOs or FIXMEs. These comments should help you bring your code to a state where it can be merged.

Geolykt committed:
Signficantly reduced the Complexity of previous fixes

The new complexity is now closer to O(1); Credits to @rfrowe for giving the idea. Also contains minor javadoc changes as well as a patch that would fix an error which renders the explosion event handlers useless

Co-Authored-By: Ryan Rowe <ryanf.rowe@gmail.com>

rfrowe commented on Zenchantments/src/main/java/zedly/zenchantments/WatcherEnchant.java:
```diff
@@ -115,55 +123,20 @@ public void onBlockPistonExtendEvent(BlockPistonExtendEvent evt) {
-        
-        return false;
+    public boolean isProtectedBlock(Block block) {
+        return NetherStep.netherstepLocs.containsKey(block.getLocation()) || FrozenStep.frozenLocs.containsKey(block.getLocation());
```
Looks great! Did you test that this works?

Geolykt replied:
yep, although only in 1.16.1, so without any NMS

rfrowe commented on Zenchantments/src/main/java/zedly/zenchantments/WatcherEnchant.java:
```diff
@@ -115,55 +123,20 @@ public void onBlockPistonExtendEvent(BlockPistonExtendEvent evt) {
-        }
-        
-        return false;
+    public boolean isProtectedBlock(Block block) {
```
Can be `private`

Geolykt replied:
maybe also make it static since it doesn't call any methods that would require it to be aware of the instance? Although this has almost no effects I guess since it's private method

Geolykt committed:
visibillity changes

maybe also make it static?

Co-Authored-By: Ryan Rowe <ryanf.rowe@gmail.com>

rfrowe reviewed (commented):
@Zedly Can we test this? Let's get it in

Geolykt commented:
Due to severe changes in the project's layout I will not be able to push these changes. I will reopen a remastered version of this PR shortly
